### DISCUSSION TITLE
Open encrypted SQLite DB in desktop editor/browser

### DISCUSSION THREAD

**Discussion by cremfert** [simolus3/drift#1479] [Q&A] [answered]
Hey everybody,

first of all, thanks @simolus3 for providing such a great library!
I'm working with Drift FFI and SQLCipher as mentioned here: https://drift.simonbinder.eu/docs/other-engines/encryption/

It seems to be working fine. If I create a database and insert some data, I can access the data after restart of the app. If I change the password in code, it - correctly - gives me an error and I cannot access the database.

However, I wanted to double check that everything is working fine. So I downloaded the .sqlite file from the Android emulator (via Android Studio) and also from the iOS simulator. Unfortunately, I cannot open the files, neither in TablePlus (Mac app), SQLiteStudio nor DB Browser for SQLite.
- TablePlus does not show a password prompt and just does nothing
- DB Browser for SQLite shows a password prompt, but if I enter my password, it reopens the password prompt again (I just tried different settings, e.g. SQLCipher3 and SQLCipher4)
- SQLiteStudio rejects opening the file and always complains that the database file is not valid.

Especially SQLiteStudio's behavior irritates me as it suggests a corrupted database file but this stands in contrast to the behavior of the app in the emulator/simulator where I can open the database.

Is this normal? Is there a way to open the DB file on the desktop?

Thanks a lot for help!

**Comment by davidmartos96** [comment, marked as answer]
@cremfert I use DB Browser. You should be able to open it with the password and the default SQLCipher 4 settings.
You can try to do the same process of creating a db on mobile, download and open it, but this time passing an empty string as password on your app (unencrypted) and see if you can open it.
Maybe you are getting an invalid file after downloading it from the mobile device.

**Reply by cremfert** [reply]
@davidmartos96 Thanks a lot for your comment. If I do `rawDb.execute("PRAGMA key = '';");` then I can open the file in DB Browser. But it is not working when the database is encrypted. Any further ideas? Do I need to update a SQLite library on my system or something like that?

Thanks a lot!

**Reply by davidmartos96** [reply]
@cremfert Yes, you should make sure that you have the right Sqlcipher binaries installed on the machine. From a terminal window you can type "sqlcipher" and see you will see the version you have installed

**Reply by cremfert** [reply]
@davidmartos96 Thanks for your help! I made a step forward: if I now copy the file from the Android emulator, I can open it - even the encrypted one - in DB Browser. I needed to setup the SQLCipher library before running `runApp(myApp())` in `main.dart`.

However, when copying the file from the iOS simulator, it is still not working. Maybe it has something to do with the file system there...

**Reply by davidmartos96** [reply]
@cremfert Ok that looks to me that it could be related to this: https://github.com/davidmartos96/sqflite_sqlcipher/blob/master/sqflite/README.md#if-using-sqflite-as-direct-or-transitive-dependency

Are you using a package like cache_network_image that relies on sqflite?

**Reply by cremfert** [reply]
@davidmartos96 Yes, indeed, I am using cached_network_image! I will check your hints in the file and will give an update here! Thanks!

**Reply by davidmartos96** [reply]
@cremfert Although I think that only happens when using the non FFI approach for SQLCipher. Although it's possible that when setting up sqlcipher on iOS, the normal SQLite library is loaded instead of SQLCipher because both are available and both share the same API

**Reply by davidmartos96** [reply]
@cremfert To be sure you can run the following query from your app and see what you get: "PRAGMA cipher_version". You should get something like "sqlcipher community 4.x"

**Reply by cremfert** [reply]
@davidmartos96 Thanks a lot, you made my day! I followed your instructions for the `dependency_override` and not it's working perfectly. `PRAGMA cipher_version` returns `{cipher_version: 4.4.3 community}` so it looks fine now. And I can open the file in DB browser.

I think, the information about the dependency is very crucial and should be mentioned in the moor documentation. I will open a ticket there.

Again, thanks a lot for your valuable help!

**Reply by cremfert** [reply]
For everybody having the same issue: The solution is mentioned here: https://github.com/davidmartos96/sqflite_sqlcipher/blob/master/sqflite/README.md#if-using-sqflite-as-direct-or-transitive-dependency

```
dependency_overrides:
  sqflite:
    git:
      url: https://www.github.com/davidmartos96/sqflite_sqlcipher.git
      path: sqflite
      ref: fmdb_override
```

All credits to @davidmartos96 for this solutions!

**Reply by cremfert** [reply]
One last question, @davidmartos96: does this `dependency_override` only affects iOS or do we also need to modify something for Android? Thanks a lot!

**Reply by davidmartos96** [reply]
@cremfert Glad that worked! As said, I think that solution is great most of the time when using the non FFI approach, but after using a while ago SQLCipher with FFI along the official Admob package, I encountered with the same issue and the override didn't work. 
The problem was thas Admob on iOS links internally with Sqlite, so the app was using the Sqlite implementation instead of Sqlcipher. 
The solution was different and it may be more appropriate, taken from this thread: https://discuss.zetetic.net/t/cannot-open-encrypted-database-with-sqlcipher-4/3654/4

In summary to link SQLCipher on your app: From Xcode -> Project -> Other Linker Flags -> "-framework SQLCipher" (without quotes)

Could you try that instead of the override solution?

cc @simolus3 I believe this is valuable information to include in the docs if it works as a general solution

**Reply by davidmartos96** [reply]
> One last question, @davidmartos96: does this `dependency_override` only affects iOS or do we also need to modify something for Android? Thanks a lot!

It only changes the pod used on iOS so that the sqflite package picks up the SQLCipher binaries

**Reply by cremfert** [reply]
> Could you try that instead of the override solution?

Yes sure, I will try that tomorrow and will write an update here!

**Reply by cremfert** [reply]
@davidmartos96 I now did some tests:
- Without the dependency_overrides and without the other linker flag options, encryption IS NOT enabled. `PRAGMA cipher_version` throws an error!
- With dependency_overrides OR other linker flag options: encryption is working and showing: `{cipher_version: 4.4.3 community}
`

I guess that this could lead to security issues when developers just "believe" the app and do not double-check the encryption, as without the changes to the pubspec.yml or the XCode project, it works on the device but actually without encryption! I will comment on the issue further.

Thanks again for your help!

**Reply by davidmartos96** [reply]
Yes, I definitely agree!
